### PR TITLE
I12243 checker

### PR DESCRIPTION
**Pull request by romani**
#12243 

idea is taken from https://github.com/checkstyle/checkstyle/pull/11617/commits/0d149b5db810c2f04da20329d30e49436a96624a

```
 groovy ./.ci/checker-framework.groovy --list
WARNING: A.........
............All illegal access operations will be denied in a future release
Supported profiles:
checker-framework-nullness-optional-interning
checker-framework-signature-gui-units-init
checker-framework-formatter
checker-framework-methods-resource-fenum
checker-framework-lock-tainting
checker-framework-index
checker-framework-regex-property-key-compiler-message
checker-framework-purity-value-returns

```

-----

example of diff if I damaged suppression file manually:
```
$ git diff
diff --git a/.ci/checker-framework-suppressions/checker-framework-lock-tainting-suppressions.xml b/.ci/checker-framework-suppressions/checker-framework-lock-tainting-suppressions.xml
index 3107aa6..5280c7f 100644
--- a/.ci/checker-framework-suppressions/checker-framework-lock-tainting-suppressions.xml
+++ b/.ci/checker-framework-suppressions/checker-framework-lock-tainting-suppressions.xml
@@ -6,7 +6,7 @@
     <message>Incompatible receiver type</message>
     <lineContent>.filter(ExternalResourceHolder.class::isInstance)</lineContent>
     <details>
-      found   : @GuardedBy Class&lt;@GuardedBy ExternalResourceHolder&gt;
+      found   : @GuardedBy Class&lt;@GuardedBy &gt;
       required: @GuardSatisfied Class&lt;@GuardedBy ExternalResourceHolder&gt;
       Consequence: method
       @GuardedBy Class&lt;@GuardedBy ExternalResourceHolder&gt;

```

### CODE REVIEW

**Review by rnveach — APPROVED** [review]

**Comment by romani** [comment]
here is small diff from referenced commit 
```
...
-  <checkerFrameworkError unstable="false">
-    <fileName>checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java</fileName>
-    <specifier>methodref.receiver</specifier>
-    <message>Incompatible receiver type</message>
-    <lineContent>.collect(BitSet::new, BitSet::set, BitSet::or);</lineContent>
-    <details>
-      found   : @GuardSatisfied BitSet
-      required: @GuardedBy BitSet
-      Consequence: method in @GuardedBy BitSet
-      void set(@GuardSatisfied BitSet this, @GuardedBy int p0)
-      is not a valid method reference for method in @GuardedBy ObjIntConsumer&lt;@GuardedBy BitSet&gt;
-      void accept(@GuardedBy ObjIntConsumer&lt;@GuardedBy BitSet&gt; this, @GuardedBy BitSet p0, @GuardedBy int p1)
-    </details>
-  </checkerFrameworkError>
-</suppressedErrors>
+[INFO] Error stacktraces are turned on.
+[INFO] Scanning for projects...
+[INFO] Inspecting build with total of 1 modules...
+[INFO] Installing Nexus Staging features:
+[INFO]   ... total of 1 executions of maven-deploy-plugin replaced with nexus-staging-maven-plugin
+[INFO] 
+[INFO] ------------------< com.puppycrawl.tools:checkstyle >-------------------
+[INFO] Building checkstyle 10.3.5-SNAPSHOT
....
```

**Comment by romani** [comment]
git diff will damage unstable in XML, so we need to do git diff only if `checker-framework.groovy` execution is failed.
this update will do a trick:

```
$ git diff
diff --git a/.github/workflows/checker-framework.yml b/.github/workflows/checker-framework.yml
index 066c4dc..1122ad1 100644
--- a/.github/workflows/checker-framework.yml
+++ b/.github/workflows/checker-framework.yml
@@ -35,5 +35,6 @@ jobs:
         run: groovy ./.ci/checker-framework.groovy ${{ matrix.profile }} -g
         env:
           MAVEN_OPTS: "-Xmx4g"
-      - name: Git Diff
+      - name: Git Diff on fail
         run: ./.ci/validation.sh git-diff
+        when: on_fail
```

File: .github/workflows/checker-framework.yml [hunk 1]
```diff
@@ -32,6 +32,9 @@ jobs:
       - name: Checkout
         uses: actions/checkout@v2
       - name: Execute checker framework
-        run: groovy ./.ci/checker-framework.groovy ${{ matrix.profile }}
+        run: groovy ./.ci/checker-framework.groovy ${{ matrix.profile }} -g
```
Comment (rnveach) [inline .github/workflows/checker-framework.yml:35]: `-g` doesn't write to file

**Review by rnveach — CHANGES_REQUESTED** [review]

**Comment by romani** [comment]
as remove columns we have duplicates
```
[WARNING] /home/rivanov/java/github/romani/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java:[339,38] [methodref.receiver] Incompatible receiver type
  found   : @GuardSatisfied BitSet
  required: @GuardedBy BitSet
  Consequence: method in @GuardedBy BitSet
    void set(@GuardSatisfied BitSet this, @GuardedBy int p0)
  is not a valid method reference for method in @GuardedBy ObjIntConsumer<@GuardedBy BitSet>
    void accept(@GuardedBy ObjIntConsumer<@GuardedBy BitSet> this, @GuardedBy BitSet p0, @GuardedBy int p1)
[WARNING] /home/rivanov/java/github/romani/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/utils/TokenUtil.java:[339,51] [methodref.receiver] Incompatible receiver type
  found   : @GuardSatisfied BitSet
  required: @GuardedBy BitSet
  Consequence: method in @GuardedBy BitSet
    void or(@GuardSatisfied BitSet this, @GuardedBy BitSet p0)
  is not a valid method reference for method in @GuardedBy BiConsumer<@GuardedBy BitSet, @GuardedBy BitSet>
    void accept(@GuardedBy BiConsumer<@GuardedBy BitSet, @GuardedBy BitSet> this, @GuardedBy BitSet p0, @GuardedBy BitSet p1)
```
but I think it is fine to keep them.

**Comment by romani** [comment]
@nrmancuso , @Vyom-Yadav , FYI.
Please share feedback

**Review by rnveach — APPROVED** [review]

**Comment by rnveach** [comment]
https://github.com/checkstyle/checkstyle/actions/runs/3223386884/jobs/5273411194
````
Summary: TotalFiles=4 FilesWithViolations=1 P1=0 P2=2 P3=3

File: checker-framework.groovy
    Violation: Rule=ExplicitArrayListInstantiation P=2 Line=165 Msg=[ArrayList objects are better instantiated using the form "[]"] Src=[final List<CheckerFrameworkError> errors = new ArrayList<>()]
    Violation: Rule=ExplicitArrayListInstantiation P=2 Line=223 Msg=[ArrayList objects are better instantiated using the form "[]"] Src=[final List<CheckerFrameworkError> suppressedErrors = new ArrayList<>()]
    Violation: Rule=UnnecessarySemicolon P=3 Line=95 Msg=[Semi-colons as line endings can be removed safely] Src=[    FileWriter writer = new FileWriter(suppressionFile);]
    Violation: Rule=UnnecessarySemicolon P=3 Line=97 Msg=[Semi-colons as line endings can be removed safely] Src=[    writer.flush();]
    Violation: Rule=UnnecessarySemicolon P=3 Line=98 Msg=[Semi-colons as line endings can be removed safely] Src=[    writer.close();]
````

File: .github/workflows/checker-framework.yml [hunk 2]
```diff
@@ -35,3 +35,6 @@ jobs:
         run: groovy ./.ci/checker-framework.groovy ${{ matrix.profile }}
         env:
           MAVEN_OPTS: "-Xmx4g"
+      - name: Git Diff on fail
+        run: ./.ci/validation.sh git-diff
+        when: on_fail
```
Comment (romani) [inline .github/workflows/checker-framework.yml:40]: This is not working as it is not circleci

**Review by romani — COMMENTED** [review]
Items

File: .github/workflows/checker-framework.yml (hunk 2 above)
Comment (romani) [reply .github/workflows/checker-framework.yml]: fixed.

File: .ci/checker-framework.groovy [hunk 3]
```diff
@@ -75,26 +68,36 @@ private static Set<String> getCheckerFrameworkProfiles() {
...
         final Node suppressedErrorsNode = xmlParser.parse(suppressedErrorsFileUri)
         suppressedErrors = getSuppressedErrors(suppressedErrorsNode)
     }
-    return compareErrors(errors, suppressedErrors, flag)
+
+    String newSuppresion = '<?xml version="1.0" encoding="UTF-8"?>\n<suppressedErrors>'
+    errors.each {
+        newSuppresion += it.toXmlString() + '\n'
+    }
+    newSuppresion += "</suppressedErrors>\n"
```
Comment (rnveach) [inline .ci/checker-framework.groovy:94]: Can we print empty line if `errors` is empty so start/end on separate lines (similar to our {}s on ifs)

**Comment by rnveach** [comment]
@romani 
https://github.com/checkstyle/checkstyle/actions/runs/3236477172/jobs/5302336199#step:6:17
> -      found   : @Initialized @Nullable Constructor&lt;capture#934 extends @Initialized @Nullable Object&gt;
> -      required: @Initialized @NonNull Constructor&lt;capture#934 extends @Initialized @Nullable Object&gt;
> +      found   : @Initialized @Nullable Constructor&lt;capture#372 extends @Initialized @Nullable Object&gt;
> +      required: @Initialized @NonNull Constructor&lt;capture#372 extends @Initialized @Nullable Object&gt;

This PR relies on https://github.com/checkstyle/checkstyle/pull/12244 .

**Comment by romani** [comment]
bunch of problems, new violations 
and
order changes https://github.com/checkstyle/checkstyle/actions/runs/3250491425/jobs/5334241891 :
```
diff --git a/.ci/checker-framework-suppressions/checker-framework-regex-property-key-compiler-message-suppressions.xml b/.ci/checker-framework-suppressions/checker-framework-regex-property-key-compiler-message-suppressions.xml
index f00592c..9884b08 100644
--- a/.ci/checker-framework-suppressions/checker-framework-regex-property-key-compiler-message-suppressions.xml
+++ b/.ci/checker-framework-suppressions/checker-framework-regex-property-key-compiler-message-suppressions.xml
@@ -1,5 +1,16 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <suppressedErrors>
+  <checkerFrameworkError unstable="false">
+    <fileName>checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/LocalizedMessage.java</fileName>
+    <specifier>argument</specifier>
+    <message>incompatible argument for parameter key of getString.</message>
+    <lineContent>final String pattern = resourceBundle.getString(key);</lineContent>
+    <details>
+      found   : @UnknownPropertyKey String
+      required: @PropertyKey String
+    </details>
+  </checkerFrameworkError>
+
   <checkerFrameworkError unstable="false">
     <fileName>checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/Main.java</fileName>
     <specifier>argument</specifier>
@@ -121,17 +132,6 @@
     </details>
   </checkerFrameworkError>
 
-  <checkerFrameworkError unstable="false">
-    <fileName>checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/LocalizedMessage.java</fileName>
-    <specifier>argument</specifier>
-    <message>incompatible argument for parameter key of getString.</message>
-    <lineContent>final String pattern = resourceBundle.getString(key);</lineContent>
-    <details>
-      found   : @UnknownPropertyKey String
-      required: @PropertyKey String
-    </details>
-  </checkerFrameworkError>
-
   <checkerFrameworkError unstable="false">
     <fileName>checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/AvoidEscapedUnicodeCharactersCheck.java</fileName>
     <specifier>argument</specifier>
```

so even order or problem catching is not deterministic, so sad.

File: .ci/checker-framework.groovy (hunk 3 above)
Comment (romani) [reply .ci/checker-framework.groovy:97]: done

**Pull request merged**